File: Training/Advanced_Cluster_Usage/Exercise1/mem_test.c
#include <stdio.h>
#include <stdlib.h>

// Function prototype for the random number generator
double ran3(int *idum);

int main() {
    int n = 60000; // Matrix dimension
    int i, j;
    int iseed = -99;
    double **h;

    // Allocate memory for the matrix
    h = (double **)malloc(n * sizeof(double *));
    for (i = 0; i < n; i++) {
        h[i] = (double *)malloc(n * sizeof(double));
    }

    // Create random symmetric test matrix
    for (i = 0; i < n; i++) {
        for (j = 0; j <= i; j++) {
            h[i][j] = ran3(&iseed);
            h[j][i] = h[i][j]; // Symmetric matrix
        }
    }
    printf("Hamiltonian matrix created successfully!\n");

    // Free memory
    for (i = 0; i < n; i++) {
        free(h[i]);
    }
    free(h);

    return 0;
}

// Random number generator (ran3)
double ran3(int *idum) {
    static int iff = 0;
    static int inext, inextp;
    static long ma[55];
    static int mbig = 1000000000, mseed = 161803398, mz = 0;
    static double fac = 1.0 / 1000000000.0;
    long mj, mk;
    int i, ii, k;

    if (*idum < 0 || iff == 0) {
        iff = 1;
        mj = mseed - (*idum < 0 ? -*idum : *idum);
        mj %= mbig;
        ma[54] = mj;
        mk = 1;
        for (i = 1; i <= 54; i++) {
            ii = (21 * i) % 55;
            ma[ii - 1] = mk;
            mk = mj - mk;
            if (mk < mz) mk += mbig;
            mj = ma[ii - 1];
        }
        for (k = 1; k <= 4; k++) {
            for (i = 0; i < 55; i++) {
                ma[i] -= ma[(i + 30) % 55];
                if (ma[i] < mz) ma[i] += mbig;
            }
        }
        inext = 0;
        inextp = 31;
        *idum = 1;
    }

    if (++inext == 55) inext = 0;
    if (++inextp == 55) inextp = 0;
    mj = ma[inext] - ma[inextp];
    if (mj < mz) mj += mbig;
    ma[inext] = mj;
    return mj * fac;
}

File: Training/Advanced_Cluster_Usage/Exercise1/Fortran/mem_test.f90
!=====================================================================
! Program: mem_test.f90
!          Program generates a symetric random matrix of dimension 60K
!=====================================================================
program mem_test
  implicit none
  integer(4) :: n = 60000 ! Matrix dimension
  integer(4) :: i
  integer(4) :: j
  integer(4) :: iseed
  real(8), allocatable :: h(:,:)
  
  ! Random number generator to compute a random test-matrix
  real(8), external :: ran3
  
  iseed = -99

  ! Allocate memory ..................................................
  if ( .not. allocated (h) ) allocate ( h(n,n) )

  ! Create random test matrix h.......................................
  do i = 1, n
     do j = 1, i
        h(i,j) = ran3(iseed)
        h(j,i) = h(i,j)
     end do
  end do
  write(6,*) 'Hamiltonian matrix created successfully!'

  ! Free memory ......................................................
  if ( allocated( h ) ) deallocate( h )

end program mem_test

!=====================================================================
!     The function
!        ran3
!     returns a uniform random number deviate between 0.0 and 1.0. Set
!     the idum to any negative value to initialize or reinitialize the
!     sequence. Any large MBIG, and any small (but still large) MSEED
!     can be substituted for the present values.
!=====================================================================
REAL(8) FUNCTION ran3(idum)
  IMPLICIT NONE
  INTEGER :: idum
  INTEGER :: mbig,mseed,mz
  REAL(8) ::  fac
  PARAMETER (mbig=1000000000,mseed=161803398,mz=0,fac=1./mbig)
  INTEGER :: i,iff,ii,inext,inextp,k
  INTEGER :: mj,mk,ma(55)
  SAVE iff,inext,inextp,ma
  DATA iff /0/

  IF ( (idum < 0) .or. (iff == 0) ) THEN
     iff=1
     mj=mseed-IABS(idum)
     mj=MOD(mj,mbig)
     ma(55)=mj
     mk=1
     DO i=1,54
        ii=MOD(21*i,55)
        ma(ii)=mk
        mk=mj-mk
        IF(mk < mz)mk=mk+mbig
        mj=ma(ii)
     ENDDO
     DO k=1,4
        DO i=1,55
           ma(i)=ma(i)-ma(1+MOD(i+30,55))
           IF (ma(i) < mz)ma(i)=ma(i)+mbig
        ENDDO
     ENDDO
     inext=0
     inextp=31
     idum=1
  ENDIF
  inext=inext+1
  IF (inext == 56) inext=1
  inextp=inextp+1
  IF (inextp == 56) inextp=1
  mj=ma(inext)-ma(inextp)
  IF (mj < mz) mj=mj+mbig
  ma(inext)=mj
  ran3=mj*fac
  return
END FUNCTION ran3
